homeassistant/components/syncthing/__init__.py

```diff
@@ -0,0 +1,168 @@
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, cancel_listen_task)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        name = entry.data[CONF_NAME]
+        await hass.data[DOMAIN][name].unsubscribe()
+        hass.data[DOMAIN].pop(name)
+
+    return unload_ok
+
+
+class SyncthingClient:
+    """A Syncthing client."""
+
+    def __init__(self, hass, client, name):
+        """Initialize the client."""
+        self._hass = hass
+        self._client = client
+        self._name = name
+        self._listen_task = None
+
+    @property
+    def database(self):
+        """Get database namespace client."""
+        return self._client.database
+
+    @property
+    def system(self):
+        """Get system namespace client."""
+        return self._client.system
+
+    def subscribe(self):
+        """Start event listener coroutine."""
+        self._listen_task = self._hass.loop.create_task(self._listen())
+
+    async def unsubscribe(self):
+        """Stop event listener coroutine."""
+        if self._listen_task:
+            self._listen_task.cancel()
+        await self._client.close()
+
+    async def _listen(self):
+        """Listen to Syncthing events."""
+        events = self._client.events
+        server_was_unavailable = False
+        while True:
+            try:
+                await self._client.system.ping()
+                if server_was_unavailable:
+                    _LOGGER.info(
+                        "The syncthing server '%s' is back online.", self._name
+                    )
+                    async_dispatcher_send(
+                        self._hass, f"{SERVER_AVAILABLE}-{self._name}"
+                    )
+                    server_was_unavailable = False
+
+                async for event in events.listen():
+                    if events.last_seen_id == 0:
+                        continue  # skipping historical events from the first batch
+                    if event["type"] not in EVENTS:
+                        continue
+
+                    signal_name = EVENTS[event["type"]]
+                    folder = None
+                    if "folder" in event["data"]:
+                        folder = event["data"]["folder"]
+                    else:  # A workaround, some events store folder id under `id` key
+                        folder = event["data"]["id"]
+                    async_dispatcher_send(
+                        self._hass,
+                        f"{signal_name}-{self._name}-{folder}",
+                        event,
+                    )
+                return
+            except aiosyncthing.exceptions.SyncthingError:
```

Comment:
Please only wrap the line that can raise in the try... except block. Use an `else:` block if needed.

Reply:
I extracted a method `_server_available()` to get around the first place one the exception can be raised.

I also had to wrap the whole `async for event in events.listen():` loop to a try-except block because the `events.listen()` async generator is the place where an exception can happen. 

I think there is no better way to squeeze the  try-except block even more.